src/main/java/Tree/RedBlackTree.java

```diff
@@ -0,0 +1,372 @@
+                    if (nodeToUpdate.getColour()) {
+                        nodeToUpdate.setDoubleBlack(true);
+                    }
+                }
+            } else {
+                swapWithPredecessor(deleteNode, predecessor);
+                nodeToUpdate = (RedBlackNode<T>) predecessor.getParentNode();
+                if (predecessor.getColour() && predecessor.getRightNode() == null && predecessor.getLeftNode() == null) {
+                    handleDoubleBlack(predecessor);
+                }
+            }
+        } else {
+            swapWithSuccessor(deleteNode, successor);
+            nodeToUpdate = (RedBlackNode<T>) successor.getParentNode();
+            if (successor.getColour() && successor.getLeftNode() == null && successor.getRightNode() == null) {
+                handleDoubleBlack(successor);
+            }
+        }
+
+        if (nodeToUpdate != null) {
+            adjustAfterDelete(nodeToUpdate);
+        }
+    }
+
+    public void adjustAfterDelete(RedBlackNode<T> node) {
+        if (node.isDoubleBlack()) {
+            handleDoubleBlack(node);
+        } else if (node.getColour()) {
+            deleteCase1(node);
+        }
+    }
+
+    public void handleDoubleBlack(RedBlackNode<T> node) {
+        RedBlackNode<T> sibling = getSibling(node);
+        RedBlackNode<T> siblingLeft = (RedBlackNode<T>) sibling.getLeftNode();
+        RedBlackNode<T> siblingRight = (RedBlackNode<T>) sibling.getRightNode();
+        RedBlackNode<T> parent = (RedBlackNode<T>) node.getParentNode();
+
+        if (sibling.getColour()) {
+            if (node.getKey() > parent.getKey()) {
+                if (siblingLeft != null && !siblingLeft.getColour()) {
+                    rightRotation(parent);
+                } else if (siblingRight != null) {
+                    leftRotation(sibling);
+                    rightRotation(parent);
+                    parent.setColour(true);
+                }
+            } else if (node.getKey() < parent.getKey()) {
+                if (siblingRight != null && !siblingRight.getColour()) {
+                    leftRotation(parent);
+                } else if (siblingLeft != null) {
+                    rightRotation(sibling);
+                    leftRotation(parent);
+                    parent.setColour(true);
+                }
+            } else {
+                if (parent.getLeftNode() == null) {
+                    leftRotation(parent);
+                } else if (parent.getRightNode() == null) {
+                    rightRotation(parent);
+                }
+            }
+        } else if (!sibling.getColour()) {
+            if (node.getKey() < parent.getKey()) {
+                leftRotation(parent);
+            } else {
+                rightRotation(parent);
+            }
+            parent.setColour(false);
+            sibling.setColour(true);
+            handleDoubleBlack(node);
+        }
+    }
+
+    public void deleteCase1(RedBlackNode<T> node) {
```

Comment:
for all deleteCase#, give more descriptive names or add comments

Reply:
why can't this be combined with `adjustAfterDelete` and `deleteCase3`?
this is just one single deleteCase